crates/iceberg-catalog/src/implementations/postgres/task_queues.rs

```diff
@@ -0,0 +1,1114 @@
+            picked_up_at: task.picked_up_at,
+            attempt: task.attempt,
+            state: task.task_data,
+        }));
+    }
+
+    Ok(None)
+}
+
+async fn record_success(
+    id: Uuid,
+    pool: &PgPool,
+    message: Option<&str>,
+) -> Result<(), IcebergErrorResponse> {
+    let _ = sqlx::query!(
+        r#"
+        WITH history as (
+            INSERT INTO task_log(task_id, warehouse_id, queue_name, task_data, status, entity_id, entity_type, message, attempt)
+                SELECT task_id, warehouse_id, queue_name, task_data, 'success', entity_id, entity_type, $2, attempt FROM task
+                                                                        WHERE task_id = $1)
+        DELETE FROM task
+        WHERE task_id = $1
+        "#,
+        id,
+        message
+    )
+    .execute(pool)
+    .await
+    .map_err(|e| e.into_error_model("failed to record task success"))?;
+    Ok(())
+}
+
+async fn record_failure(
+    conn: &PgPool,
+    id: Uuid,
+    n_retries: i32,
+    details: &str,
+) -> Result<(), IcebergErrorResponse> {
+    let should_fail = sqlx::query_scalar!(
+        r#"
+        SELECT attempt >= $1 as "should_fail!"
+        FROM task
+        WHERE task_id = $2
+        "#,
+        n_retries,
+        id
+    )
+    .fetch_optional(conn)
+    .await
+    .map_err(|e| e.into_error_model("failed to check if task should fail"))?
+    .unwrap_or(false);
+
+    if should_fail {
+        sqlx::query!(
+            r#"
+            WITH history as (
+                INSERT INTO task_log(task_id, warehouse_id, queue_name, task_data, status, entity_id, entity_type, attempt)
+                SELECT task_id, warehouse_id, queue_name, task_data, 'failure', entity_id, entity_type, attempt
+                FROM task WHERE task_id = $1
+            )
+            DELETE FROM task
+            WHERE task_id = $1
+            "#,
+            id
+        )
+        .execute(conn)
+        .await
+        .map_err(|e| e.into_error_model("failed to log and delete failed task"))?;
+    } else {
+        sqlx::query!(
+            r#"
+            WITH task_log as (
+                INSERT INTO task_log(task_id, warehouse_id, queue_name, task_data, status, entity_id, entity_type, message, attempt)
+                SELECT task_id, warehouse_id, queue_name, task_data, 'failure', entity_id, entity_type, $2, attempt
+                FROM task WHERE task_id = $1
+            )
+            UPDATE task
+            SET status = 'pending'
+            WHERE task_id = $1
+            "#,
+            id,
+            details
+        )
+        .execute(conn)
+        .await
+        .map_err(|e| e.into_error_model("failed to update task status"))?;
+    }
+
+    Ok(())
+}
+
+use crate::service::task_queue::{EntityId, TaskInput, TaskMetadata};
+
+/// Cancel pending tasks for a warehouse
+/// If `task_ids` are provided in `filter` which are not pending, they are ignored
+pub(crate) async fn cancel_pending_tasks(
```

Comment:
Shouldn't we also cancel running tasks via this function?
My idea is (doesn't need to be this PR):
* User requests cancellation (are there other reasons to cancel currently?) -> set status `cancellation_requested`
* Running task can periodically check its own status via a trait method. If it sees `cancellation_requested` state, it cancels its run and `record_cancellation`
What do you think?
If the task doesn't check cancellation, then the task will stay in `cancellation_requested` until max age is reached.